docs/development/synapse_architecture/streams.md

```diff
@@ -0,0 +1,126 @@
+## Streams
+
+Synapse has a concept of "streams", which are roughly described in [`id_generators.py`](
+    https://github.com/matrix-org/synapse/blob/develop/synapse/storage/util/id_generators.py
+).
+It is important to understand this, so let's describe them formally.
+We paraphrase from the docstring of [`AbstractStreamIdGenerator`](
+    https://github.com/matrix-org/synapse/blob/a719b703d9bd0dade2565ddcad0e2f3a7a9d4c37/synapse/storage/util/id_generators.py#L96
+).
+
+A stream is an append-only log `T1, T2, ..., Tn, ...` of facts which grows over time.
+Only "writers" can add facts to a stream, and there may be multiple writers.
+
+Each fact has an ID, called its "stream ID".
+Readers should only process facts in ascending stream ID order.
+
+Roughly speaking, each stream is backed by a database table.
+It should have a `stream_id` column holding stream IDs, plus additional columns
+as necessary to describe the fact.
+(Note that it may take multiple rows (with the same `stream_id`) to describe that fact.)
+Stream IDs are globally unique (enforced by Postgres sequences).
```

Comment:
> Does "globally unique" mean across all types of streams or within a particular stream table?

Within a given stream. Thanks, that's an important oversight on my part.

> How can there be multiple rows with the same stream_id but also have them be unique?

Err, good question. I think I should have said something like

> Within a stream, no two facts may have the same `stream_id`.
> (Note that it may take multiple rows to describe a single fact; if so, each of those rows will share a stream ID.)

I think this needs an example to drive it home.